### PR TITLE
Fix config 64 max limit

### PR DESCRIPTION
**Pull request by Arimodu**
Stop using WaitHandle and use an action instead

### CODE REVIEW

**Comment by Arimodu** [comment]
Fixes #87

**Comment by Meivyn** [comment]
You should clean up your PR. You have extra spaces, you're not respecting access modifiers conventions, and you have unused variables. And that's what I see from the diff only. Just my two cents on it.

**Comment by nike4613** [comment]
This changes behavior in a way which defeats the point. This implementation causes saving to be done on the thread which made the change, significantly slowing anything that changes the config, and obviating the save thread alltogether. I would rather manually implement multi-wait than this.

There's also the compatibility consideration-- this changes a public interface *intended* to external consumers to be able to implement, which is a breaking change. (I'm not sure how many people actually use it, but it is still a binary breaking change.)

**Comment by Arimodu** [comment]
It does not (or should not) do that. It is a nested action. The outer action (the thread that made the change) just enqueues itself into the queue, and then the save thread created processes the blocking queue one by one.

As for the breaking change, thats true, I tried to make it work without changing but it either introduced a delay or didnt save to disk unless the game was restarted.

File: IPA.Loader/Config/ConfigRuntime.cs [hunk 1]
```diff
@@ -31,7 +26,12 @@ public int GetHashCode(DirectoryInfo obj)
         }
 
         private static readonly ConcurrentBag<Config> configs = new ConcurrentBag<Config>();
-        private static readonly AutoResetEvent configsChangedWatcher = new AutoResetEvent(false);
+        private static readonly Action configsChangedAction = () =>
+        {
+            foreach (var config in configs.Where(c => c.Store != null).ToArray())
+                config.Store.SyncAction = () => RequiresSave.Add(() => Save(config));
```
Comment (Arimodu) [inline IPA.Loader/Config/ConfigRuntime.cs]: Note here,
`() => RequiresSave.Add()` is the outer action and
`() => Save(config)` is the inner action that is going to be processed by the save thread

**Comment by Arimodu** [comment]
Ok, I went back and redone the whole thing. I believe this way it could work for you?
Instead of changing the interface I went and directly referenced the static collection, adding directly from the implementation. Then I just check if the reference to the WriteSyncObject is the same and call `Save()` on that.
I also left the old implementation in tact, just adding another .Where that checks for null on the SyncObject, this way, the old implementation only handles those where the SyncObject is not null (which it is set to from the new implementation)

File: IPA.Loader/Config/Stores/GeneratedStoreImpl/MakeCreator.cs [hunk 2]
```diff
@@ -430,15 +430,18 @@ private static (GeneratedStoreCreator ctor, Type type) MakeCreator(Type type)
...
             var IConfigStore_WriteTo = IConfigStore_t.GetMethod(nameof(IConfigStore.WriteTo));
             var IConfigStore_ReadFrom = IConfigStore_t.GetMethod(nameof(IConfigStore.ReadFrom));
 
             #region IConfigStore.SyncObject
-            var syncObjProp = typeBuilder.DefineProperty(nameof(IConfigStore.SyncObject), PropertyAttributes.None, IConfigStore_GetSyncObject.ReturnType, null);
-            var syncObjPropGet = typeBuilder.DefineMethod($"<g>{nameof(IConfigStore.SyncObject)}", virtualPropertyMethodAttr, syncObjProp.PropertyType, Type.EmptyTypes);
+            var syncObjProp = typeBuilder.DefineProperty(nameof(IConfigStore.SyncAction), PropertyAttributes.None, IConfigStore_GetSyncObject.ReturnType, null);
+            var syncObjPropGet = typeBuilder.DefineMethod($"get_{nameof(IConfigStore.SyncAction)}", virtualPropertyMethodAttr, syncObjProp.PropertyType, Type.EmptyTypes);
+            var syncObjPropSet = typeBuilder.DefineMethod($"set_{nameof(IConfigStore.SyncAction)}", virtualPropertyMethodAttr, null, new[] { syncObjProp.PropertyType });
```
Comment (nike4613) [inline IPA.Loader/Config/Stores/GeneratedStoreImpl/MakeCreator.cs]: This looks like a remnant of the old implementation. Shouldn't this be reverted?

Comment (Arimodu) [reply IPA.Loader/Config/Stores/GeneratedStoreImpl/MakeCreator.cs]: Oops, you are right, totally missed that while reverting...

**Comment by Arimodu** [comment]
Ok reverted, one last thing, what should I put as the deprecation message on the SyncObject?

**Comment by nike4613** [comment]
Something to the effect of "this is no longer the preferred mechanism for requesting a save". One other thing I noticed is that the new request system is not at all usable by manually implemented providers. (This isn't necessarily a problem, but *is* an argument in favor of not actually deprecating the wait handle, but instead just noting that for internally-implemented providers it may not be triggered).

**Comment by Arimodu** [comment]
I see, thats true, the class is internal.... However, couldnt that be solved with a public facing method? Understandable if that is not preffered from your side, but personally I dont see a problem.

**Comment by Arimodu** [comment]
Looking at it again, I dont see a straightforward way of implementing it....

One way could be to make a public wrapper class that could contain both the (internal / private) collection and a method that would add to it - this seems overly complicated to me

Another way would be to make `ConfigRuntime` public, internalize all methods inside except the one we need to add to this save thread collection - This seems ok to me, however I am not sure if that is what you want to do

**Comment by nike4613** [comment]
I'm honestly OK with not exposing it, and just keeping the old `WaitHandle` approach functional for such implementers.

**Comment by Arimodu** [comment]
Alright then, I'm gonna push the notice about the internally implemented waithandle and I think this is ready for merging

**Pull request merged**